crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -546,6 +546,9 @@ struct ScopeInferenceExtra<'db> {
 
     /// The diagnostics for this region.
     diagnostics: TypeCheckDiagnostics,
+
+    /// The inferred signature of a function definition (without any default values filled in).
+    signature: Option<InferredFunctionSignature<'db>>,
```

Comment:
I suspect that this is one of the main reasons for the memory increase and slowdown. The idea of extra is that we only create it in rare instances. However, assuming I understand your change correctly, we now create an extra for every function definition of which there are many. 

Could we maybe use a type inference context instead?